#### McDonalds EDA


### 1. Problem Statement:

- McDonald's is hip all over the world, including in India. However, the McDonald's menu is not universal. In India, there are some menu items that are similar, but others that are much different from the U.S. menu. Through comparing the McDonald's menu in the U.S. and India, we can see that fast food menus can truly give us a perspective on cultural and economic similarities and differences throughout the world.



### 2. Importing Packages

In [1]:
import pandas as pd 
import numpy as np 
pd.set_option('mode.chained_assignment',None)
pd.set_option('display.max_colwidth',-1)
pd.options.display.max_columns = 50
import warnings
warnings.filterwarnings('ignore')

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_11536\1510049603.py:4: FutureWarning: Passing a negative integer is deprecated in version 1.0 and will not be supported in future version. Instead, use None to not limit the column width.
  pd.set_option('display.max_colwidth',-1)


In [2]:
import matplotlib.pyplot as plt 
%matplotlib inline 

import seaborn as sns
sns.set(style='whitegrid',font_scale=1.3,color_codes=True)


### 3. Loading Data

In [3]:
df_menu = pd.read_excel("C:\\Users\\Lenovo\\Desktop\\Project\\Mcd's EDA\\mcdonalds_menu.xlsx")
df_menu.head()

Category                  Item Serve_Size  Energy  Protein  Total_Fat  \
0  Breakfast  Egg & Cheese Muffin     112        290     14       13.0        
1  Breakfast  Sausage McMuffm        112        273     16       11.0        
2  Breakfast  Sausage & Egg McMuffm  157        355     22       17.0        
3  Breakfast  Veg McMuffm            119        299     10       12.0        
4  Breakfast  Veg Supreme Muffin      139        299     7        13.0        

   Saturated_Fat  Trans_Fat  Cholestrol  Carbohydrates  Sugars  Dietary_Fibre  \
0  7.0            0.2        244         28             2       2               
1  5.7            0.2        50          28             2       2               
2  7.9            0.2        277         29             2       2               
3  7.2            0.2        21          38             3       3               
4  5.2            0.2        11          39             4       4               

  Sodium  
0  620    
1  950    
2  1020   
3  1000   
4  960

In [4]:
pip install chart_studio

Note: you may need to restart the kernel to use updated packages.


In [5]:
from plotly.offline import init_notebook_mode, iplot
import plotly.graph_objs as go
import chart_studio.plotly as py
from plotly import tools
init_notebook_mode(connected=True)

In [6]:
df_menu.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 13 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Category       340 non-null    object 
 1   Item           340 non-null    object 
 2   Serve_Size     340 non-null    object 
 3   Energy         340 non-null    int64  
 4   Protein        340 non-null    int64  
 5   Total_Fat      340 non-null    float64
 6   Saturated_Fat  340 non-null    float64
 7   Trans_Fat      340 non-null    float64
 8   Cholestrol     340 non-null    int64  
 9   Carbohydrates  340 non-null    int64  
 10  Sugars         340 non-null    int64  
 11  Dietary_Fibre  340 non-null    int64  
 12  Sodium         340 non-null    object 
dtypes: float64(3), int64(6), object(4)
memory usage: 34.7+ KB


In [7]:
# Importing 2nd dataset
df_store = pd.read_excel(r"C:\Users\Lenovo\Desktop\Project\Mcd's EDA\mcdonalds_outlets.xlsx")
df_store.head()

Brand      Store_ID     Store_Name Ownership_Type  \
0  McDonald's  23149-228271  Banjara Hills  Joint Venture   
1  McDonald's  23191-228548  Kukatpally     Joint Venture   
2  McDonald's  23193-228546  Madhapur       Joint Venture   
3  McDonald's  23180-228545  Jubilee Hills  Joint Venture   
4  McDonald's  24457-238129  Hi-Tech City   Joint Venture   

                                                                  Street_Address  \
0  Lower Ground Floor, GVK One, Road Number 1, Banjara Hills                       
1  Upper Ground Floor, Forum Sujana Mall, Kukatpally                               
2  Lower Ground Floor, Inorbit Mall, Madhapur                                      
3  Ground Floor, Road No. 92, Near Apollo hospital, Jubilee Hills                  
4  Upper Ground Floor, Phoenix tower A, Opposite Trident Hotel, Madhapur Village   

        City State Country  Postcode Phone_Number                  Timezone  \
0  Hyderabad  AP    IN      500034    NaN          GMT+05:30 Asia/New_Delhi   
1  Hyderabad  AP    IN      500072    NaN          GMT+05:30 Asia/New_Delhi   
2  Hyderabad  AP    IN      500081    NaN          GMT+05:30 Asia/New_Delhi   
3  Hyderabad  AP    IN      500033    NaN          GMT+05:30 Asia/New_Delhi   
4  Hyderabad  AP    IN      500084    NaN          GMT+05:30 Asia/New_Delhi   

   Longitude  Latitude   Revenue   Profits  Selling_Price  Cost_Price  \
0  78.45      17.42     2.117344  0.171584  1.655165       0.907433     
1  78.39      17.48     1.058504  0.054645  0.895944       0.453645     
2  78.39      17.43     4.505020  0.663867  2.864311       1.930723     
3  78.42      17.42     3.405530  0.397835  2.394072       1.459513     
4  78.38      17.45     7.332713  1.656524  3.395053       3.142591     

   Gross_Profit_Margin  Number_of_Employees     Customers  \
0  0.747732             34.311197            3979.583117    
1  0.442299             25.487533            1156.010620    
2  0.933588             54.208502            10346.720786   
3  0.934559             45.046083            7414.746544    
4  0.252461             77.772607            17887.234107   

      Best_Selling_Item  
0  Egg & Cheese Muffin     
1  Sausage McMuffm        
2  Sausage & Egg McMuffm  
3  Veg McMuffm            
4  Veg Supreme Muffin

In [8]:
df_store.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 21 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Brand                340 non-null    object 
 1   Store_ID             340 non-null    object 
 2   Store_Name           340 non-null    object 
 3   Ownership_Type       340 non-null    object 
 4   Street_Address       340 non-null    object 
 5   City                 340 non-null    object 
 6   State                340 non-null    object 
 7   Country              340 non-null    object 
 8   Postcode             340 non-null    int64  
 9   Phone_Number         248 non-null    object 
 10  Timezone             340 non-null    object 
 11  Longitude            340 non-null    float64
 12  Latitude             340 non-null    float64
 13  Revenue              340 non-null    float64
 14  Profits              340 non-null    float64
 15  Selling_Price        340 non-null    flo

In [9]:
df_store['Best_Selling_Item'].nunique()

340

In [10]:
df_merge = pd.merge(df_store, df_menu, how='inner', left_on='Best_Selling_Item',right_on='Item').drop(['Item'],1)
df_merge.head()

Brand      Store_ID     Store_Name Ownership_Type  \
0  McDonald's  23149-228271  Banjara Hills  Joint Venture   
1  McDonald's  23191-228548  Kukatpally     Joint Venture   
2  McDonald's  23193-228546  Madhapur       Joint Venture   
3  McDonald's  23180-228545  Jubilee Hills  Joint Venture   
4  McDonald's  24457-238129  Hi-Tech City   Joint Venture   

                                                                  Street_Address  \
0  Lower Ground Floor, GVK One, Road Number 1, Banjara Hills                       
1  Upper Ground Floor, Forum Sujana Mall, Kukatpally                               
2  Lower Ground Floor, Inorbit Mall, Madhapur                                      
3  Ground Floor, Road No. 92, Near Apollo hospital, Jubilee Hills                  
4  Upper Ground Floor, Phoenix tower A, Opposite Trident Hotel, Madhapur Village   

        City State Country  Postcode Phone_Number                  Timezone  \
0  Hyderabad  AP    IN      500034    NaN          GMT+05:30 Asia/New_Delhi   
1  Hyderabad  AP    IN      500072    NaN          GMT+05:30 Asia/New_Delhi   
2  Hyderabad  AP    IN      500081    NaN          GMT+05:30 Asia/New_Delhi   
3  Hyderabad  AP    IN      500033    NaN          GMT+05:30 Asia/New_Delhi   
4  Hyderabad  AP    IN      500084    NaN          GMT+05:30 Asia/New_Delhi   

   Longitude  Latitude   Revenue   Profits  Selling_Price  Cost_Price  \
0  78.45      17.42     2.117344  0.171584  1.655165       0.907433     
1  78.39      17.48     1.058504  0.054645  0.895944       0.453645     
2  78.39      17.43     4.505020  0.663867  2.864311       1.930723     
3  78.42      17.42     3.405530  0.397835  2.394072       1.459513     
4  78.38      17.45     7.332713  1.656524  3.395053       3.142591     

   Gross_Profit_Margin  Number_of_Employees     Customers  \
0  0.747732             34.311197            3979.583117    
1  0.442299             25.487533            1156.010620    
2  0.933588             54.208502            10346.720786   
3  0.934559             45.046083            7414.746544    
4  0.252461             77.772607            17887.234107   

      Best_Selling_Item   Category Serve_Size  Energy  Protein  Total_Fat  \
0  Egg & Cheese Muffin     Breakfast  112        290     14       13.0        
1  Sausage McMuffm        Breakfast  112        273     16       11.0        
2  Sausage & Egg McMuffm  Breakfast  157        355     22       17.0        
3  Veg McMuffm            Breakfast  119        299     10       12.0        
4  Veg Supreme Muffin      Breakfast  139        299     7        13.0        

   Saturated_Fat  Trans_Fat  Cholestrol  Carbohydrates  Sugars  Dietary_Fibre  \
0  7.0            0.2        244         28             2       2               
1  5.7            0.2        50          28             2       2               
2  7.9            0.2        277         29             2       2               
3  7.2            0.2        21          38             3       3               
4  5.2            0.2        11          39             4       4               

  Sodium  
0  620    
1  950    
2  1020   
3  1000   
4  960

In [11]:
df_merge.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 340 entries, 0 to 339
Data columns (total 33 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Brand                340 non-null    object 
 1   Store_ID             340 non-null    object 
 2   Store_Name           340 non-null    object 
 3   Ownership_Type       340 non-null    object 
 4   Street_Address       340 non-null    object 
 5   City                 340 non-null    object 
 6   State                340 non-null    object 
 7   Country              340 non-null    object 
 8   Postcode             340 non-null    int64  
 9   Phone_Number         248 non-null    object 
 10  Timezone             340 non-null    object 
 11  Longitude            340 non-null    float64
 12  Latitude             340 non-null    float64
 13  Revenue              340 non-null    float64
 14  Profits              340 non-null    float64
 15  Selling_Price        340 non-null    flo

- There are a total of 340 samples (rows) and 33 columns in the dataframe.

- There are 19 columns with a numeric datatype and 14 columns with an object datatype.

- There are missing values in the Phone Number column.

In [12]:
df_merge.describe().round(2)

Postcode  Longitude  Latitude  Revenue  Profits  Selling_Price  \
count  3.400000e+02  340.00     340.00    340.00   340.00   340.00          
mean   2.902141e+08 -56.76      33.49     25.85    4.64     14.60           
std    3.649787e+08  76.97      9.95      14.48    3.41     9.47            
min    2.134000e+03 -158.02     12.91     1.00     0.05     0.85            
25%    9.043625e+04 -99.32      28.52     15.30    1.83     6.91            
50%    5.600790e+05 -87.26      35.10     27.06    3.89     13.71           
75%    6.604731e+08 -71.12      40.10     37.93    6.92     20.43           
max    9.966980e+08  80.26      61.60     49.68    13.51    40.28           

       Cost_Price  Gross_Profit_Margin  Number_of_Employees  Customers  \
count  340.00      340.00               340.00               340.00      
mean   11.08       3.52                 90.56                13071.99    
std    6.20        5.19                 31.35                6993.91     
min    0.43       -4.88                 25.01                1002.93     
25%    6.56        0.07                 64.58                6877.50     
50%    11.60       1.43                 89.16                13373.55    
75%    16.26       6.18                 115.52               19245.19    
max    21.29       19.46                149.09               24964.84    

        Energy  Protein  Total_Fat  Saturated_Fat  Trans_Fat  Cholestrol  \
count  340.00   340.00   340.00     340.00         340.00     340.00       
mean   337.80   11.69    13.05      5.60           0.18       47.54        
std    231.85   10.99    13.64      5.22           0.38       81.82        
min    0.00     0.00     0.00       0.00           0.00       0.00         
25%    176.50   2.00     1.38       0.50           0.00       4.00         
50%    299.50   10.00    10.00      4.55           0.00       25.00        
75%    460.00   17.00    20.00      9.00           0.10       55.00        
max    1880.00  87.00    118.00     24.10          2.50       575.00       

       Carbohydrates  Sugars  Dietary_Fibre  
count  340.00         340.00  340.00         
mean   43.77          26.35   1.54           
std    26.94          26.90   1.66           
min    0.00           0.00    0.00           
25%    28.00          4.00    0.00           
50%    41.00          16.00   1.00           
75%    56.00          43.25   3.00           
max    141.00         128.00  9.00

### 3.2 Pandas Profiling before Data Preprocessing

In [13]:
from pydantic_settings import BaseSettings
!pip install pydantic==1.10.11

  Obtaining dependency information for pydantic==1.10.11 from https://files.pythonhosted.org/packages/05/2c/09a35e6c16e206c80906ce93c0dd1e86d8b9738ea05eb82fe38c8f0cfab1/pydantic-1.10.11-cp310-cp310-win_amd64.whl.metadata
  Using cached pydantic-1.10.11-cp310-cp310-win_amd64.whl.metadata (150 kB)
Using cached pydantic-1.10.11-cp310-cp310-win_amd64.whl (2.1 MB)
  Attempting uninstall: pydantic
    Found existing installation: pydantic 2.6.1
    Uninstalling pydantic-2.6.1:
      Successfully uninstalled pydantic-2.6.1


ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
pydantic-settings 2.1.0 requires pydantic>=2.3.0, but you have pydantic 1.10.11 which is incompatible.
ydata-profiling 4.6.4 requires pydantic>=2, but you have pydantic 1.10.11 which is incompatible.


In [14]:
import ydata_profiling
from ydata_profiling import ProfileReport

ModuleNotFoundError: No module named 'pydantic.v1'

In [ ]:
profile = profile = ydata_profiling.ProfileReport(df_merge)

profile.to_file(output_file="MCD_profiling_before_preprocessing.html")

Summarize dataset:   0%|          | 0/5 [00:00<?, ?it/s]

Generate report structure:   0%|          | 0/1 [00:00<?, ?it/s]

Render HTML:   0%|          | 0/1 [00:00<?, ?it/s]

Export report to file:   0%|          | 0/1 [00:00<?, ?it/s]

### Observations from Pandas Profiling before Data Processing

- Dataset info:

Number of variables: 33,
Number of observations: 340,
Missing cells: 92 (0.8%),

- Variables types:

Numeric = 16,
Categorical = 10,
Text (Unique) = 3,
Rejected = 4,

### 4.Data Preprocessing

- Here we will perform Data Preprocessing on our dataset, to make the data usable for EDA.

- In order to avoid any errors, please run Pandas Profiling before running the code below.



In [ ]:
df_merge.head()

Brand      Store_ID     Store_Name Ownership_Type  \
0  McDonald's  23149-228271  Banjara Hills  Joint Venture   
1  McDonald's  23191-228548  Kukatpally     Joint Venture   
2  McDonald's  23193-228546  Madhapur       Joint Venture   
3  McDonald's  23180-228545  Jubilee Hills  Joint Venture   
4  McDonald's  24457-238129  Hi-Tech City   Joint Venture   

                                                                  Street_Address  \
0  Lower Ground Floor, GVK One, Road Number 1, Banjara Hills                       
1  Upper Ground Floor, Forum Sujana Mall, Kukatpally                               
2  Lower Ground Floor, Inorbit Mall, Madhapur                                      
3  Ground Floor, Road No. 92, Near Apollo hospital, Jubilee Hills                  
4  Upper Ground Floor, Phoenix tower A, Opposite Trident Hotel, Madhapur Village   

        City State Country  Postcode Phone_Number                  Timezone  \
0  Hyderabad  AP    IN      500034    NaN          GMT+05:30 Asia/New_Delhi   
1  Hyderabad  AP    IN      500072    NaN          GMT+05:30 Asia/New_Delhi   
2  Hyderabad  AP    IN      500081    NaN          GMT+05:30 Asia/New_Delhi   
3  Hyderabad  AP    IN      500033    NaN          GMT+05:30 Asia/New_Delhi   
4  Hyderabad  AP    IN      500084    NaN          GMT+05:30 Asia/New_Delhi   

   Longitude  Latitude   Revenue   Profits  Selling_Price  Cost_Price  \
0  78.45      17.42     2.117344  0.171584  1.655165       0.907433     
1  78.39      17.48     1.058504  0.054645  0.895944       0.453645     
2  78.39      17.43     4.505020  0.663867  2.864311       1.930723     
3  78.42      17.42     3.405530  0.397835  2.394072       1.459513     
4  78.38      17.45     7.332713  1.656524  3.395053       3.142591     

   Gross_Profit_Margin  Number_of_Employees     Customers  \
0  0.747732             34.311197            3979.583117    
1  0.442299             25.487533            1156.010620    
2  0.933588             54.208502            10346.720786   
3  0.934559             45.046083            7414.746544    
4  0.252461             77.772607            17887.234107   

      Best_Selling_Item   Category Serve_Size  Energy  Protein  Total_Fat  \
0  Egg & Cheese Muffin     Breakfast  112        290     14       13.0        
1  Sausage McMuffm        Breakfast  112        273     16       11.0        
2  Sausage & Egg McMuffm  Breakfast  157        355     22       17.0        
3  Veg McMuffm            Breakfast  119        299     10       12.0        
4  Veg Supreme Muffin      Breakfast  139        299     7        13.0        

   Saturated_Fat  Trans_Fat  Cholestrol  Carbohydrates  Sugars  Dietary_Fibre  \
0  7.0            0.2        244         28             2       2               
1  5.7            0.2        50          28             2       2               
2  7.9            0.2        277         29             2       2               
3  7.2            0.2        21          38             3       3               
4  5.2            0.2        11          39             4       4               

  Sodium  
0  620    
1  950    
2  1020   
3  1000   
4  960

- Since the column Brand has a constant value for all the samples, we will drop this column.

- Also the columns Postcode, Phone Number, and Timezone will be dropped because they don't provide any additional insights into the data.

In [ ]:
df_merge.drop(['Brand','Postcode','Phone_Number','Timezone'],axis = 1,inplace = True)

In [ ]:
df_merge.head()

Store_ID     Store_Name Ownership_Type  \
0  23149-228271  Banjara Hills  Joint Venture   
1  23191-228548  Kukatpally     Joint Venture   
2  23193-228546  Madhapur       Joint Venture   
3  23180-228545  Jubilee Hills  Joint Venture   
4  24457-238129  Hi-Tech City   Joint Venture   

                                                                  Street_Address  \
0  Lower Ground Floor, GVK One, Road Number 1, Banjara Hills                       
1  Upper Ground Floor, Forum Sujana Mall, Kukatpally                               
2  Lower Ground Floor, Inorbit Mall, Madhapur                                      
3  Ground Floor, Road No. 92, Near Apollo hospital, Jubilee Hills                  
4  Upper Ground Floor, Phoenix tower A, Opposite Trident Hotel, Madhapur Village   

        City State Country  Longitude  Latitude   Revenue   Profits  \
0  Hyderabad  AP    IN      78.45      17.42     2.117344  0.171584   
1  Hyderabad  AP    IN      78.39      17.48     1.058504  0.054645   
2  Hyderabad  AP    IN      78.39      17.43     4.505020  0.663867   
3  Hyderabad  AP    IN      78.42      17.42     3.405530  0.397835   
4  Hyderabad  AP    IN      78.38      17.45     7.332713  1.656524   

   Selling_Price  Cost_Price  Gross_Profit_Margin  Number_of_Employees  \
0  1.655165       0.907433    0.747732             34.311197             
1  0.895944       0.453645    0.442299             25.487533             
2  2.864311       1.930723    0.933588             54.208502             
3  2.394072       1.459513    0.934559             45.046083             
4  3.395053       3.142591    0.252461             77.772607             

      Customers     Best_Selling_Item   Category Serve_Size  Energy  Protein  \
0  3979.583117   Egg & Cheese Muffin     Breakfast  112        290     14        
1  1156.010620   Sausage McMuffm        Breakfast  112        273     16        
2  10346.720786  Sausage & Egg McMuffm  Breakfast  157        355     22        
3  7414.746544   Veg McMuffm            Breakfast  119        299     10        
4  17887.234107  Veg Supreme Muffin      Breakfast  139        299     7         

   Total_Fat  Saturated_Fat  Trans_Fat  Cholestrol  Carbohydrates  Sugars  \
0  13.0       7.0            0.2        244         28             2        
1  11.0       5.7            0.2        50          28             2        
2  17.0       7.9            0.2        277         29             2        
3  12.0       7.2            0.2        21          38             3        
4  13.0       5.2            0.2        11          39             4        

   Dietary_Fibre Sodium  
0  2              620    
1  2              950    
2  2              1020   
3  3              1000   
4  4              960

#### Rounding the values of Revenue, Profits, Sellling_Price, Cost_Price, Gross_Profit_Margin columns to 3 decimal places.

In [ ]:
df_merge = df_merge.round(decimals = 2)
df_merge.head()

Store_ID     Store_Name Ownership_Type  \
0  23149-228271  Banjara Hills  Joint Venture   
1  23191-228548  Kukatpally     Joint Venture   
2  23193-228546  Madhapur       Joint Venture   
3  23180-228545  Jubilee Hills  Joint Venture   
4  24457-238129  Hi-Tech City   Joint Venture   

                                                                  Street_Address  \
0  Lower Ground Floor, GVK One, Road Number 1, Banjara Hills                       
1  Upper Ground Floor, Forum Sujana Mall, Kukatpally                               
2  Lower Ground Floor, Inorbit Mall, Madhapur                                      
3  Ground Floor, Road No. 92, Near Apollo hospital, Jubilee Hills                  
4  Upper Ground Floor, Phoenix tower A, Opposite Trident Hotel, Madhapur Village   

        City State Country  Longitude  Latitude  Revenue  Profits  \
0  Hyderabad  AP    IN      78.45      17.42     2.12     0.17      
1  Hyderabad  AP    IN      78.39      17.48     1.06     0.05      
2  Hyderabad  AP    IN      78.39      17.43     4.51     0.66      
3  Hyderabad  AP    IN      78.42      17.42     3.41     0.40      
4  Hyderabad  AP    IN      78.38      17.45     7.33     1.66      

   Selling_Price  Cost_Price  Gross_Profit_Margin  Number_of_Employees  \
0  1.66           0.91        0.75                 34.31                 
1  0.90           0.45        0.44                 25.49                 
2  2.86           1.93        0.93                 54.21                 
3  2.39           1.46        0.93                 45.05                 
4  3.40           3.14        0.25                 77.77                 

   Customers     Best_Selling_Item   Category Serve_Size  Energy  Protein  \
0  3979.58    Egg & Cheese Muffin     Breakfast  112        290     14        
1  1156.01    Sausage McMuffm        Breakfast  112        273     16        
2  10346.72   Sausage & Egg McMuffm  Breakfast  157        355     22        
3  7414.75    Veg McMuffm            Breakfast  119        299     10        
4  17887.23   Veg Supreme Muffin      Breakfast  139        299     7         

   Total_Fat  Saturated_Fat  Trans_Fat  Cholestrol  Carbohydrates  Sugars  \
0  13.0       7.0            0.2        244         28             2        
1  11.0       5.7            0.2        50          28             2        
2  17.0       7.9            0.2        277         29             2        
3  12.0       7.2            0.2        21          38             3        
4  13.0       5.2            0.2        11          39             4        

   Dietary_Fibre Sodium  
0  2              620    
1  2              950    
2  2              1020   
3  3              1000   
4  4              960

In [ ]:
# Using apply (np.floor) to get the floor values.

df_merge['Number_of_Employees'] = df_merge['Number_of_Employees'].apply(np.floor).astype(int)
df_merge['Customers'] = df_merge['Customers'].apply(np.floor).astype(int)
df_merge.head()

Store_ID     Store_Name Ownership_Type  \
0  23149-228271  Banjara Hills  Joint Venture   
1  23191-228548  Kukatpally     Joint Venture   
2  23193-228546  Madhapur       Joint Venture   
3  23180-228545  Jubilee Hills  Joint Venture   
4  24457-238129  Hi-Tech City   Joint Venture   

                                                                  Street_Address  \
0  Lower Ground Floor, GVK One, Road Number 1, Banjara Hills                       
1  Upper Ground Floor, Forum Sujana Mall, Kukatpally                               
2  Lower Ground Floor, Inorbit Mall, Madhapur                                      
3  Ground Floor, Road No. 92, Near Apollo hospital, Jubilee Hills                  
4  Upper Ground Floor, Phoenix tower A, Opposite Trident Hotel, Madhapur Village   

        City State Country  Longitude  Latitude  Revenue  Profits  \
0  Hyderabad  AP    IN      78.45      17.42     2.12     0.17      
1  Hyderabad  AP    IN      78.39      17.48     1.06     0.05      
2  Hyderabad  AP    IN      78.39      17.43     4.51     0.66      
3  Hyderabad  AP    IN      78.42      17.42     3.41     0.40      
4  Hyderabad  AP    IN      78.38      17.45     7.33     1.66      

   Selling_Price  Cost_Price  Gross_Profit_Margin  Number_of_Employees  \
0  1.66           0.91        0.75                 34                    
1  0.90           0.45        0.44                 25                    
2  2.86           1.93        0.93                 54                    
3  2.39           1.46        0.93                 45                    
4  3.40           3.14        0.25                 77                    

   Customers     Best_Selling_Item   Category Serve_Size  Energy  Protein  \
0  3979       Egg & Cheese Muffin     Breakfast  112        290     14        
1  1156       Sausage McMuffm        Breakfast  112        273     16        
2  10346      Sausage & Egg McMuffm  Breakfast  157        355     22        
3  7414       Veg McMuffm            Breakfast  119        299     10        
4  17887      Veg Supreme Muffin      Breakfast  139        299     7         

   Total_Fat  Saturated_Fat  Trans_Fat  Cholestrol  Carbohydrates  Sugars  \
0  13.0       7.0            0.2        244         28             2        
1  11.0       5.7            0.2        50          28             2        
2  17.0       7.9            0.2        277         29             2        
3  12.0       7.2            0.2        21          38             3        
4  13.0       5.2            0.2        11          39             4        

   Dietary_Fibre Sodium  Number_of_Employess  
0  2              620    34                   
1  2              950    25                   
2  2              1020   54                   
3  3              1000   45                   
4  4              960    77

In [ ]:
# Serve Size has some irregularities
# Checking the differernt types of values in the Serve_Size coloumn 
df_merge['Serve_Size'].unique()

array([112, 157, 119, 139, '142 g', 64, 115, 246, '173 g', '136 g', 130,
       '169 g', 146, 199, 272, '186 g', 260, 110, 156, '220 g', 87, 96,
       144, 90, 180, 9, 25, 14, 10, 200, 275, 402, 277, 400, '249 g', 344,
       494, 249, '344 g', 300, '500 g', 250, 237, 247, 81, 93, '97 g',
       '92 g', '132 g', 92, 132, 111, 155, 147, 94, 167, '200 g', 141,
       133, 121, '123 g', 136, 135, 161, 185, 150, 164, 153, 117, 131,
       163, 177, 181, 143, 201, 174, 178, 205, 197, 241, '269 g', 283,
       '286 g', 420, 434, '423 g', 437, 151, 192, '111 g', 114, 251,
       '211 g', 202, 227, 235, 244, 98, 113, 270, '190 g', 159, 208, 213,
       230, 217, '284 g', 160, 171, 148, 316, 302, 310, 297, 314, 318,
       304, 291, 65, 97, 162, 323, 646, 142, 223, 255, 348, 335, '130 g',
       123, 116, 128, '121 g', 75, 168, '38 g', 34, 149, 77, '33 g', 33,
       29, 179, 182, 453, 595, 850, 340, '453 g', 236, 170, 623, 479,
       '595 g', 566, '566 g', '340 g', 907, '907 g', '623 g', 460

In [ ]:
df_merge['Serve_Size'] = df_merge['Serve_Size'].astype(str).str.replace("'","").str.replace(" g","")

In [ ]:
df_merge['Serve_Size'].unique()

array(['112', '157', '119', '139', '142', '64', '115', '246', '173',
       '136', '130', '169', '146', '199', '272', '186', '260', '110',
       '156', '220', '87', '96', '144', '90', '180', '9', '25', '14',
       '10', '200', '275', '402', '277', '400', '249', '344', '494',
       '300', '500', '250', '237', '247', '81', '93', '97', '92', '132',
       '111', '155', '147', '94', '167', '141', '133', '121', '123',
       '135', '161', '185', '150', '164', '153', '117', '131', '163',
       '177', '181', '143', '201', '174', '178', '205', '197', '241',
       '269', '283', '286', '420', '434', '423', '437', '151', '192',
       '114', '251', '211', '202', '227', '235', '244', '98', '113',
       '270', '190', '159', '208', '213', '230', '217', '284', '160',
       '171', '148', '316', '302', '310', '297', '314', '318', '304',
       '291', '65', '162', '323', '646', '223', '255', '348', '335',
       '116', '128', '75', '168', '38', '34', '149', '77', '33', '29',
       '179', '182', 

In [ ]:
df_merge['Serve_Size']= df_merge['Serve_Size'].astype(int)
df_merge['Serve_Size'].unique()

array([112, 157, 119, 139, 142,  64, 115, 246, 173, 136, 130, 169, 146,
       199, 272, 186, 260, 110, 156, 220,  87,  96, 144,  90, 180,   9,
        25,  14,  10, 200, 275, 402, 277, 400, 249, 344, 494, 300, 500,
       250, 237, 247,  81,  93,  97,  92, 132, 111, 155, 147,  94, 167,
       141, 133, 121, 123, 135, 161, 185, 150, 164, 153, 117, 131, 163,
       177, 181, 143, 201, 174, 178, 205, 197, 241, 269, 283, 286, 420,
       434, 423, 437, 151, 192, 114, 251, 211, 202, 227, 235, 244,  98,
       113, 270, 190, 159, 208, 213, 230, 217, 284, 160, 171, 148, 316,
       302, 310, 297, 314, 318, 304, 291,  65, 162, 323, 646, 223, 255,
       348, 335, 116, 128,  75, 168,  38,  34, 149,  77,  33,  29, 179,
       182, 453, 595, 850, 340, 236, 170, 623, 479, 566, 907, 460, 207,
       285, 381, 403])

In [ ]:
df_merge['Serve_Size'].dtype

dtype('int32')

In [ ]:
df_merge['Sodium'].value_counts()[:5]

-      15
0      14
180    11
150    8 
140    8 
Name: Sodium, dtype: int64

In [15]:
# Calculate the median of the rows which do not have sodium value.

df_merge[df_merge['Sodium'] == '-']['Best_Selling_Item']

38    Regular Coca-Cola   
39    Medium Coca-Cola    
40    Large Coca-Cola     
41    Regular Fanta Orange
42    Medium Fanta Orange 
43    Large Fanta Orange  
44    Regular Thums-up    
45    Medium Thums-up     
46    Large Thums - up    
47    Regular Sprite      
48    Medium Sprite       
49    Large Sprite        
50    Diet Coke           
73    Black Tea           
75    Black coffee         
Name: Best_Selling_Item, dtype: object

In [17]:
df_merge[~(df_merge['Sodium']=='-')]['Sodium'].median()

230.0